aws_lambda_powertools/event_handler/appsync.py

```diff
@@ -149,7 +149,12 @@ def lambda_handler(event, context):
             Router.current_event = data_model(event)
             response = self._call_single_resolver(event=event, data_model=data_model)
 
-        self.clear_context()
+        # We don't clear the context for coroutines because we don't have control over the event loop.
+        # If we clean the context immediately, it might not be available when the coroutine is actually executed.
+        # For single async operations, the context should be cleaned up manually after the coroutine completes.
+        # See: https://github.com/aws-powertools/powertools-lambda-python/issues/5290
+        if not asyncio.iscoroutine(response):
```

Comment:
As discussed during our sync, before going forward with this implementation I think we could explore calling `self.clear_context()` at the beginning of this function rather than at the end (i.e. ~L141).

My reasoning here is that since the point of this call is to clear the context after each request so that no info is leaked to the next one, then we could move this logic at the beginning and achieve the same outcome.

Since we are explicitly targeting AWS Lambda in this feature, we can assume that we have a single entry point and that all it matters is that at the beginning of a request the context is clean. Cleaning it before ensures this but also removes the issue of awaiting async operations entirely.

Overall, at least based on my mental model of the utility (which might be incorrect), it doesn't matter if we leave the context dirty after a request as long as it's cleaned before the next request since Lambda might recycle the environment anyway.

Let me know if you think this is feasible at all, ultimately I don't have enough context on the full implementation and I might be completely off base.

Reply:
Hi @dreamorosi! I've spent some time investigating this and unfortunately there is no solution for this other than refactoring the way we run the single resolvers in async mode. If I refactor this, I can introduce a major breaking change for those who have control of the async/eventloop event, as I would have to make Powertools have this control and I have no idea of ​​the consequences for those who are using it the current way.

So, I suggest we move on with the current solution.

Reply:
Ok, makes sense.

Not sure if you want to track this for a future version, but either way we can resolve the discussion and move forward for now.

Reply:
Yep, good point! I'm adding a comment "# TODO" to track that. I'll need a new approval.